### PR TITLE
Issue #19259 Fix locking during log file rotation and make rotateByCo…

### PR DESCRIPTION
**Pull request by mikehaertl**
…py mandatory

| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ✔️
| Fixed issues  | #19259

### CODE REVIEW

File: framework/log/FileTarget.php [hunk 1]
```diff
@@ -117,31 +122,21 @@ public function export()
             clearstatcache();
         }
         if ($this->enableRotation && @filesize($this->logFile) > $this->maxFileSize * 1024) {
-            @flock($fp, LOCK_UN);
-            @fclose($fp);
             $this->rotateFiles();
-            $writeResult = @file_put_contents($this->logFile, $text, FILE_APPEND | LOCK_EX);
```
Comment (mikehaertl) [inline framework/log/FileTarget.php:123]: @bizley I have mixed feelings here as I've removed something that I don't understand. Why not use `fwrite()` in both cases? I assume there were some problems caused by the messed up locking that someone tried to fix this way.

Comment (mikehaertl) [reply framework/log/FileTarget.php:123]: Ok, it dates actually back to the early days of yii2:

https://github.com/yiisoft/yii2/blame/2646e2cdcc6a260c30c8ed24928e2c9a4754d3f5/framework/log/FileTarget.php

There it did not use `ftruncate()` and so the actual logfile did no longer exist after rotation. This logic could have been removed when we introduced `ftruncate()` to clear the original logfile. It should be safe to remove it now.

**Review by bizley — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

**Comment by mikehaertl** [comment]
Do we need more live testing before merging? Locking is tricky and it's too easy to overlook something.

**Comment by bizley** [comment]
It would be great to have if possible.

**Comment by mikehaertl** [comment]
I can not really help here. Maybe @stu2162583 can try it?

File: framework/log/FileTarget.php [hunk 2]
```diff
@@ -60,15 +63,17 @@ class FileTarget extends Target
...
      * a bit faster.
      *
      * The problem with windows systems where the [rename()](https://www.php.net/manual/en/function.rename.php)
      * function does not work with files that are opened by some process is described in a
      * [comment by Martin Pelletier](https://www.php.net/manual/en/function.rename.php#102274) in
      * the PHP documentation. By setting rotateByCopy to `true` you can work
-     * around this problem.
+     * around this.
+     * @deprecated since 2.0.46 and setting it to false has no effect anymore
+     * since rotating is now always done by copy
```
Comment (samdark) [inline framework/log/FileTarget.php:76]:
```suggestion
     * since rotating is now always done by copy.
```

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Tested it on Windows 11 with concurrent console commands.

**Comment by stu2162583** [comment]
> I can not really help here. Maybe @stu2162583 can try it?

A late reply, and I sincerely apologize. At the time, I was caught up with my new job and saw the email but couldn’t respond promptly. Later, it slipped my mind. Thank you so much for fixing the bug.

Unfortunately, my new job no longer involves PHP, so I won’t be able to test this anymore.

**Pull request merged**